Dart/src/com/jetbrains/lang/dart/ide/findUsages/DartServerFindUsagesHandler.java

```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright 2000-2015 JetBrains s.r.o.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.jetbrains.lang.dart.ide.findUsages;
+
+import com.intellij.find.findUsages.FindUsagesHandler;
+import com.intellij.find.findUsages.FindUsagesOptions;
+import com.intellij.openapi.application.ApplicationManager;
+import com.intellij.openapi.application.ReadActionProcessor;
+import com.intellij.openapi.vfs.LocalFileSystem;
+import com.intellij.openapi.vfs.VirtualFile;
+import com.intellij.psi.PsiElement;
+import com.intellij.psi.PsiFile;
+import com.intellij.usageView.UsageInfo;
+import com.intellij.util.Processor;
+import com.jetbrains.lang.dart.analyzer.DartAnalysisServerService;
+import org.dartlang.analysis.server.protocol.Location;
+import org.dartlang.analysis.server.protocol.SearchResult;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+public class DartServerFindUsagesHandler extends FindUsagesHandler {
+  public DartServerFindUsagesHandler(PsiElement element) {
+    super(element);
+  }
+
+  @Override
+  public boolean processElementUsages(@NotNull final PsiElement element,
+                                      @NotNull final Processor<UsageInfo> processor,
+                                      @NotNull final FindUsagesOptions options) {
+    final ReadActionProcessor<SearchResult> searchResultProcessor = new ReadActionProcessor<SearchResult>() {
+      @Override
+      public boolean processInReadAction(SearchResult result) {
+        final Location location = result.getLocation();
+        final String filePath = location.getFile();
+        final PsiFile psiFile = findPsiFile(element, filePath);
+        if (psiFile != null) {
+          final int offset = location.getOffset();
+          final int length = location.getLength();
+          processor.process(new UsageInfo(psiFile, offset, offset + length));
+        }
+        return true;
+      }
+    };
+    // Send the search request and listen for results.
+    ApplicationManager.getApplication().runReadAction(new Runnable() {
```

Comment:
ReadAction surrounding a long operation that has wait() inside may lead to deadlock or unresponsive UI. It's better to use multiple short read actions where it is needed instead of a long one for the whole method.